### PR TITLE
Fix: Remove hashing from leaf keys in SMT Node

### PR DESCRIPTION
**Pull request by bvrooman**
The current SMT implementation hashes keys provided by calls to `update` and `delete`. The [SMT spec](https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/data_structures.md#sparse-merkle-tree) mandates leaves use the raw 32 byte key input, rather than the hash of the key input. This means clients are able to decide on the algorithm that produces 32 byte keys. 

This behaviour is an artifact of the Celestia SMT implementation. Both the Celestia implementation and the Fuel implementation are moving towards an SMT interface that takes in 32 byte keys for `update` and `delete` functions.

In this PR, I am changing the internal SMT node structure so that leaves are constructed using raw 32 byte keys. The [WIP SMT PR](https://github.com/FuelLabs/fuel-merkle/pull/41) has been similarly modified to expose the correct `update` and `delete` interface using 32 byte keys. I am also including minor changes to the Node structure from the SMT PR in this PR.

### CODE REVIEW

File: src/sparse/node.rs [hunk 1]
```diff
@@ -30,21 +30,22 @@ pub(crate) struct Node {
...
         node.set_bytes_hi(&sum(data));
         node
     }
 
     pub fn create_node(left_child: &Node, right_child: &Node) -> Self {
         let buffer = Self::default_buffer();
         let mut node = Self { buffer };
-        node.set_height(left_child.height() + 1);
-        node.set_bytes_prefix(&[NODE]);
+        let height = std::cmp::max(left_child.height(), right_child.height()) + 1;
```
Comment (adlerjohn) [inline src/sparse/node.rs:46]: This looks like an orthogonal change?

Comment (bvrooman) [reply src/sparse/node.rs:46]: Yes I may undo this change (and a couple other changes) here, but it was a necessary change in the SMT PR (I just brought the code over directly).

File: src/sparse/node.rs [hunk 2]
```diff
@@ -281,12 +286,14 @@ impl fmt::Debug for Node {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         if self.is_node() {
             f.debug_struct("Node (Internal)")
+                .field("Height", &self.height())
```
Comment (adlerjohn) [inline src/sparse/node.rs:289]: Is this necessary?

File: src/sparse/node.rs [hunk 3]
```diff
@@ -281,12 +286,14 @@ impl fmt::Debug for Node {
...
                 .field("Hash", &hex::encode(self.hash()))
                 .field("Left child key", &hex::encode(self.left_child_key()))
                 .field("Right child key", &hex::encode(self.right_child_key()))
                 .finish()
         } else {
             f.debug_struct("Node (Leaf)")
+                .field("Height", &self.height())
```
Comment (adlerjohn) [inline src/sparse/node.rs:296]: Ditto

File: src/sparse/node.rs [hunk 4]
```diff
@@ -335,6 +342,9 @@ where
     pub fn left_child(&self) -> Option<Self> {
         assert!(self.is_node());
         let key = self.node.left_child_key();
+        if key == zero_sum() {
+            return Some(Self::new(self.storage, Node::create_placeholder()));
+        }
```
Comment (adlerjohn) [inline src/sparse/node.rs:347]: Ditto, looks like an orthogonal change

File: src/sparse/node.rs [hunk 5]
```diff
@@ -343,8 +353,11 @@ where
     }
 
     pub fn right_child(&self) -> Option<Self> {
-        assert!(self.node.is_node());
+        assert!(self.is_node());
         let key = self.node.right_child_key();
+        if key == zero_sum() {
+            return Some(Self::new(self.storage, Node::create_placeholder()));
+        }
```
Comment (adlerjohn) [inline src/sparse/node.rs:360]: Ditto

**Comment by bvrooman** [comment]
I think the moral of the story is not to mix orthogonal changes
